csrc/preseg_passes/allocation_order_inference.cpp

```diff
@@ -15,442 +16,279 @@ namespace nvfuser::preseg_passes {
-  auto* out = dynamic_cast<TensorView*>(op->out());
-  if (out == nullptr) {
-    return;
-  }
-  propagateAllocationOrder(resolveAllocationOrder(op->inputs()), out);
-}
-
-void AllocationOrderInferencer::handle(PadOp* op) {
-  auto* out = dynamic_cast<TensorView*>(op->out());
-  auto* in = dynamic_cast<TensorView*>(op->in());
-  // Note: `out` from pad has rfactor domain that cannot be mapped back to
-  // `in`'s root domain. Hence we use `out`'s root domain to match permutation.
-  propagateAllocationOrder(in, out, out->getRootDomain());
-}
+  // propagate new allocation domain on dsts
+  for (TensorView* dst : dsts) {
+    // safe check when allocation domain on the entry cannot be safely mutated.
+    if (dst == nullptr || dst->hasAllocation() ||
+        fusion->getOutputAlias(dst).type != AllocationType::New) {
+      continue;
+    }
 
-void AllocationOrderInferencer::handle(ReductionOp* op) {
-  auto* out = dynamic_cast<TensorView*>(op->out());
-  auto* in = dynamic_cast<TensorView*>(op->in());
-  propagateAllocationOrder(in, out);
-}
+    // skip entry with self mapping.
+    if (hasSelfMapping(dst, exact_graph).has_value()) {
+      continue;
+    }
 
-} // namespace
+    // find a ref among srcs to be propagated to given dst
+    TensorView* ref = nullptr;
 
-// Note [ Allocation Order Propagation ]
-//
-// The propagation tries to propagate allocation order from inputs to the entire
-// fusion:
-//   1. Iterates through all inputs, looking for TensorView with allocation
-//   domain that's a permutation of its corresponding rfactor domain and record
-//   it as the allocation order of the tensor;
-//   2. Traverse the fusion IR, propagate allocation order and record results in
-//   alloc_order_map.
-std::unordered_map<const TensorView*, AllocationOrder> inferenceAllocationOrder(
-    Fusion* fusion) {
-  std::unordered_map<const TensorView*, AllocationOrder> alloc_order_map;
+    // high water mark for candidate of ref.
+    size_t non_bc_high_water_mark = 0;
+    for (const auto& iter : loop_iter_count) {
+      // discard srcs for propagation which dst has no dependency on.
+      if (!DependencyCheck::isDependencyOf(iter.first, dst)) {
+        continue;
+      }
+      // discard srcs with lower iterdomain count than ref
+      if (iter.second < non_bc_high_water_mark) {
+        // TODO: if loop_iter_count is sorted, we can early return here.
+        continue;
+      }
+      // new candidate found, update ref and high water mark
+      if (iter.second > non_bc_high_water_mark) {
+        ref = iter.first;
+        non_bc_high_water_mark = iter.second;
+        continue;
+      }
+      // found multiple candidate with the same iterdomain count
+      if (iter.second == non_bc_high_water_mark && ref != nullptr) {
+        // ensure that there's no ambiguity on permutation mapping from multiple
+        // references. we need both ref candidates to have the same mapping on
```

Comment:
Is this new logic? How do you compare this with picking an arbitrary `ref` among them when there's a tie like this.

Reply:
The old logic does propagation only through binary/ternary op, which I think is relatively easy to justify when we make an arbitrary choice.

With IdModel mapping, we are not following how the operands are *joined* together. I'm a bit hesitant to arbitrarily break the tie and I lean towards being on the conservative side.